### PR TITLE
User onboarding

### PR DESCRIPTION
**Pull request by kntsoriano**
## Description
Before the user proceeds to use our site, they must first accept the terms of service and the privacy policy.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -48,6 +48,9 @@ gem 'devise'
 # Omniauth
 gem 'omniauth-github'
 
+# Turbolinks form errors compatibility
+gem 'turbolinks_render'
```
Comment (kntsoriano) [inline Gemfile:52]: This is needed to let turbolinks support this piece of code:

```
  def create
    @onboarding = UserOnboardingForm.new(
      user_onboarding_params.merge(user: current_user)
    )

    if @onboarding.save
      redirect_to after_sign_in_path_for(current_user)
    else
      render :show
    end
  end
```

Without the gem, the `render` method won't work.

File: app/controllers/user_onboarding_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+class UserOnboardingController < ApplicationController
+  skip_before_action :ensure_onboarded!
+
+  before_action :authenticate_user!
```
Comment (iHiD) [inline app/controllers/user_onboarding_controller.rb]:
```suggestion
```

This is in ApplicationController already.

File: app/forms/user_onboarding_form.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+  validates :terms_of_service, acceptance: true
+  validates :privacy_policy, acceptance: true
+
+  def save
+    return unless valid?
+
+    user.update!(accepted_privacy_policy_at: Time.current, accepted_terms_at: Time.current)
+  end
+end
```
Comment (iHiD) [inline app/forms/user_onboarding_form.rb:14]: What's the advantage of using this form to updating the user directly? I can't see where it's adding extra value.

File: db/migrate/20201127032403_add_accept_fields_to_user.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class AddAcceptFieldsToUser < ActiveRecord::Migration[6.1]
+  def change
+    add_column :users, :accepted_privacy_policy_at, :datetime
+    add_column :users, :accepted_terms_at, :datetime
+  end
+end
```
Comment (iHiD) [inline db/migrate/20201127032403_add_accept_fields_to_user.rb]: Add these to the original create_user migration pls.

File: test/application_system_test_case.rb [hunk 5]
```diff
@@ -3,21 +3,20 @@
...
-  # driven_by :selenium, using: :chrome, screen_size: [1400, 1400]
-  driven_by :selenium, using: :headless_chrome do |driver_option|
-    # Without this argument, Chrome cannot be started in Docker
-    driver_option.add_argument('no-sandbox')
-  end
+  driven_by :selenium, using: :chrome, screen_size: [1400, 1400]
+  # driven_by :selenium, using: :headless_chrome do |driver_option|
+  #   # Without this argument, Chrome cannot be started in Docker
+  #   driver_option.add_argument('no-sandbox')
+  # end
```
Comment (iHiD) [inline test/application_system_test_case.rb]: Why this change? This always causes me pain as I end up with Chrome doing stuff visibly, not invisibly. If there is a specific test that needs this, could we just do it for the one test maybe?

File: test/controllers/tracks/exercises_controller_test.rb [hunk 6]
```diff
@@ -12,7 +12,7 @@ class Tracks::ExercisesControllerTest < ActionDispatch::IntegrationTest
   end
 
   test "index: renders correctly for joined" do
-    user = create :user
+    user = create :user, :onboarded
```
Comment (iHiD) [inline test/controllers/tracks/exercises_controller_test.rb]: I think we should presume that all users are onboarded and set these to be default in the factory. Then if we **don't** want someone to be onboarded (the edge-case) have a trait for `:not_onboarded`. 

It's worth having done it this way to check everything is right first, but if you could invert it now, I'd appreciate it 💙

**Comment by iHiD** [comment]
Looks good mate 🙂  I've made a couple of small comments.

File: app/forms/user_onboarding_form.rb (hunk 3 above)
Comment (kntsoriano) [reply app/forms/user_onboarding_form.rb:14]: There's a validation to make sure that the users checked the checkbox. We can remove that if you don't want that. Also, it allows us to return an error message related to the validation just as how you save a normal AR record.

File: test/application_system_test_case.rb (hunk 5 above)
Comment (kntsoriano) [reply test/application_system_test_case.rb]: This wasn't intended. I will revert this.

File: db/migrate/20201127032403_add_accept_fields_to_user.rb (hunk 4 above)
Comment (iHiD) [reply db/migrate/20201127032403_add_accept_fields_to_user.rb]: @kntsoriano I think you missed this one.

**Comment by iHiD** [comment]
Sweet. Merging this into the main devise branch. Thanks dude :)

**Comment by SleeplessByte** [comment]
For future updates: it's not a bad idea to store the "policy version" that was accepted. Yes, it can likely be derived from the accepted_at date, given no rolling-updates, but it's still good practice. I usually hardcode a date I wrote the file as a version identifier and store that.

**Comment by iHiD** [comment]
Yeah - we discussed this at the launch of v2 actually. I think we'd store the git_sha of the repo if we were to do it now. Thanks for the reminder.
